test/test_dataloader.py

```diff
@@ -3673,7 +3724,1264 @@ def test_out_of_order_iterable_ds(self):
+        )
+
+        # Consume some batches and track expected results
+        batches_before_checkpoint = []
+        it = iter(dl)
+        for i in range(3):  # consume first 3 batches
+            batch = next(it)
+            batches_before_checkpoint.append(batch)
+
+        # Save state after consuming 3 batches
+        state_dict = dl.state_dict()
+        self.assertIsInstance(state_dict, dict)
+
+        # Continue with original iterator and collect remaining batches
+        remaining_batches_original = []
+        for batch in it:
+            remaining_batches_original.append(batch)
+
+        # Create new loader and resume from checkpoint
+        dataset2 = StatefulMapDataset(50, shuffle=False)  # same dataset
+        dl2 = DataLoader(
+            dataset=dataset2,
+            num_workers=num_workers,
+            collate_fn=identity_collate,
+            stateful=True,
+            batch_size=5,
+        )
+        dl2.load_state_dict(state_dict)
+
+        # Collect all batches from resumed loader
+        remaining_batches_resumed = []
+        for batch in dl2:
+            remaining_batches_resumed.append(batch)
+
+        # Verify that resumed loader continues exactly where original left off
+        self.assertEqual(
+            len(remaining_batches_original), len(remaining_batches_resumed)
+        )
+        for orig, resumed in zip(remaining_batches_original, remaining_batches_resumed):
+            self.assertEqual(len(orig), len(resumed))
+            for o, r in zip(orig, resumed):
+                self.assertEqual(o, r)
+
+
+class StatefulSampler(torch.utils.data.Sampler):
+    """Sampler with stateful behavior for testing."""
+
+    def __init__(self, size):
+        self.size = size
+        self.i = 0
+
+    def __iter__(self):
+        return StatefulSamplerIterator(self.size, self.i)
+
+    def __len__(self):
+        return self.size
+
+    def state_dict(self):
+        return {"i": self.i}
+
+    def load_state_dict(self, state_dict):
+        self.i = state_dict["i"]
+
+
+class StatefulSamplerIterator:
+    """Iterator for stateful sampler."""
+
+    def __init__(self, size, start_idx=0):
+        self.size = size
+        self.i = start_idx
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        idx = self.i
+        if idx >= self.size:
+            raise StopIteration
+        self.i += 1
+        return idx
+
+    def state_dict(self):
+        return {"i": self.i}
+
+    def load_state_dict(self, state_dict):
+        self.i = state_dict["i"]
+
+
+@unittest.skipIf(
+    TEST_WITH_TSAN,
+    "Fails with TSAN with the following error: starting new threads after multi-threaded "
+    "fork is not supported. Dying (set die_after_fork=0 to override)",
+)
+class TestStatefulDataLoaderSampler(TestCase):
+    """Test stateful DataLoader with stateful samplers."""
+
+    @parametrize("num_workers", [0])
+    def test_stateful_sampler(self, num_workers):
```

Comment:
In this test you should assert that the sampler's `state_dict` and `load_state_dict` were called, and `__next__` is *not* called after load, i.e. it actually restored state and didn't fall back to fast-forward